musicbot/billboard/Billboard.py

```diff
@@ -0,0 +1,174 @@
+import asyncio
+from datetime import datetime, date
+import discord
+import logging
+import os
+from pathlib import Path
+
+from musicbot.exceptions import HelpfulError
+from musicbot.lib.JsonFile import JsonFile
+from .Song import Song
+from .SongManager import SongManager
+
+
+log = logging.getLogger(__name__)
+
+class Billboard:
+
+    def __init__(self, args):
+        self._guild_id = str(args["guild_id"])
+        self._billboard_name = args["name"]
+        self._number_of_songs_to_display = args["number_of_songs_to_display"]
+        self._number_of_days_before_recalculation_should_be_performed = args["number_of_days_before_recalculation_should_be_performed"]
+        
+        log.debug("Creating Billboard. args: {}".format(
+            str(args)
+        ))
+
+        self._base_working_directory = "data\\" + self._guild_id + "\\"
+        self._song_manager = SongManager()
+        self._billboard_content = {}
+        self._top_songs_cache = []
+        
+        self._throwHelpfulErrorIfWorkingDirectoryDoesNotExist()
+        self._initializeWorkingDirectories()
+        self._loadBillboardFile()
+
+    def _throwHelpfulErrorIfWorkingDirectoryDoesNotExist(self):
+        if not Path(self._base_working_directory).exists():
+            raise HelpfulError("Unable to proceed. Cannot find directory: {} ".format(str(self._base_working_directory)))
+        
+    def _initializeWorkingDirectories(self):
+        self._working_directory = self._base_working_directory + "billboard\\" + self._billboard_name + "\\"
+        self._song_working_directory = self._working_directory + "songs\\"
+
+        log.debug("_initializeWorkingDirectories called. working_directory: {}, song_working_dictionary: {}".format(
+            self._working_directory,
+            self._song_working_directory
+        ))
+        
+        Path(self._working_directory).mkdir(exist_ok=True)
+        Path(self._song_working_directory).mkdir(exist_ok=True)
+        
+    def _loadBillboardFile(self):
```

Comment:
You might want to look into player.py __json__ and _deserialize

Reply:
i can't recommend using the musicbot's Serializer and Serializable over our JsonFile class

the JsonFile class only requires the programmer the json_file_path (with the optional default_content) and abstracts any other details as to how this json is read and stored. its role is clear in that the programmer is to expect that it has parsed the json file into a dictionary-style data, and can pass the same type of data when saving

with the Serializer and Serialized classes its role is to recreate the object with the serialized data, but comes with the following drawbacks:
- the programmer is supposed to implement multitudes of methods, some object and some class methods. what's worse is there are no errors thrown when starting the bot that all methods have been implemented. we only find that out at runtime.
- the programmer will still need to provide these classes the file needed to be read or written (evident on how the "open" file methods are outside the scope of these classes). 
- the "json" module is also *half* exposed to the programmer, meaning the classes uses "json.dumps" to save a json file but the programmer can use any other module to load a json file.